smack-websocket/src/main/java/org/jivesoftware/smack/websocket/implementations/okhttp/OkHttpWebsocketImpl.java

```diff
@@ -0,0 +1,190 @@
+                        closeInititatedByServer = true;
+                    }
+                }
+                String closingStream = "</stream>";
+                switch (phase) {
+                case openFrameSent:
+                    if (isOpenElement(text)) {
+                        // Converts the <open> element received into <stream> element.
+                        openStreamHeader = getStreamFromOpenElement(text);
+                        phase = WebsocketConnectionPhase.readyToReceiveFeatures;
+                        try {
+                            connectionInternal.onStreamOpen(PacketParserUtils.getParserFor(openStreamHeader));
+                        } catch (XmlPullParserException | IOException e) {
+                            LOGGER.log(Level.WARNING, "Exception caught: ", e);
+                        }
+                    } else {
+                        LOGGER.log(Level.WARNING, "Unexpected Frame received", text);
+                    }
+                    break;
+                case readyToReceiveFeatures:
+                    phase = WebsocketConnectionPhase.exchangingTopLevelStreamElements;
+                    connectionInternal.parseAndProcessElement(openStreamHeader + text + closingStream);
+                    break;
+                case exchangingTopLevelStreamElements:
+                    if (isSuccessElement(text)) {
+                        phase = WebsocketConnectionPhase.readyToSendOpen;
+                    }
+                    connectionInternal.parseAndProcessElement(openStreamHeader + text + closingStream);
+                    break;
+                case closingHandshake:
+                    if (isCloseElement(text)) {
+                        connectionInternal.onStreamClosed();
+                    }
+                    break;
+                default:
+                    LOGGER.log(Level.INFO, "Default text: " + text);
+                }
+            }
+            @Override
+            public void onFailure(WebSocket webSocket, Throwable t, Response response) {
+                failureMap.put(endpoint, t);
+                LOGGER.log(Level.INFO, "Exception catched", t);
+            }
+        };
+
+        // Creates an instance of websocket through okHttpClient.
+        currentWebsocket = okHttpClient.newWebSocket(request, listener);
+
+        synchronized (connectionInternal) {
+            // Wait until websocket connection is open or either until a connection failure is detected.
+            connectionInternal.waitForCondition(() -> phase == WebsocketConnectionPhase.readyToSendOpen || failureMap.containsKey(endpoint) , "Wait for Open frame or an exception if something goes wrong");
+            if (failureMap.containsKey(endpoint)) {
+                throw failureMap.get(endpoint);
+            }
+        }
+    }
+
+    @Override
+    public void send(TopLevelStreamElement element) {
+        if (element.equals(StreamClose.INSTANCE)) {
+            disconnect();
```

Comment:
That appears wrong. The `send()` method should not bother with connection logic. It simply sends whatever it was told to send.